File: yangxj96-common/src/main/java/com/yangxj96/saas/common/respond/R.java
package com.yangxj96.saas.common.respond;

import lombok.Data;

import java.io.Serializable;

/**
 * 通用响应
 */
@Data
class R<T> implements Serializable {

    private Integer code;

    private String msg;

    private T data;


    public R(Integer code, String msg) {
        this.code = code;
        this.msg = msg;
    }

    public R(Integer code, String msg, T data) {
        this.code = code;
        this.msg = msg;
        this.data = data;
    }


    private static R<Object> success() {
        return new R<>(RStatus.SUCCESS.getCode(), RStatus.SUCCESS.getMsg());
    }

    private static <T> R<T> success(T data) {
        return new R<>(RStatus.SUCCESS.getCode(), RStatus.SUCCESS.getMsg(), data);
    }

    private static R<Object> failure() {
        return new R<>(RStatus.FAILURE.getCode(), RStatus.FAILURE.getMsg());
    }

    private static R<Object> failure(RStatus status) {
        return new R<>(status.getCode(), status.getMsg());
    }

    private static R<Object> failure(String message) {
        return new R<>(RStatus.FAILURE.getCode(), message);
    }

    private static <T> R<T> failure(RStatus status, T data) {
        return new R<>(status.getCode(), status.getMsg(), data);
    }

}

File: yangxj96-common/src/main/kotlin/com/yangxj96/saas/common/respond/R.kt
package com.yangxj96.saas.common.respond

import java.io.Serializable

/**
 * 通用响应
 */
class R<T> : Serializable {

    var code: Int? = null

    var msg: String? = null

    var data: T? = null

    constructor()

    constructor(code: Int, msg: String) {
        this.code = code
        this.msg = msg
    }

    constructor(code: Int, msg: String, data: T?) : this(code, msg) {
        this.data = data
    }



    companion object {

        fun success(): R<Any> {
            return R(RStatus.SUCCESS.code, RStatus.SUCCESS.msg)
        }

        fun <T> success(data: T?): R<T> {
            return R(RStatus.SUCCESS.code, RStatus.SUCCESS.msg, data)
        }

        fun failure(): R<Any> {
            return R(RStatus.FAILURE.code, RStatus.FAILURE.msg)
        }

        fun failure(status: RStatus): R<Any> {
            return R(status.code, status.msg)
        }

        fun failure(message: String): R<Any> {
            return R(RStatus.FAILURE.code, message)
        }

        fun <T> failure(status: RStatus, data: T?): R<T> {
            return R(status.code, status.msg, data)
        }
    }

    override fun toString(): String {
        return "R(code=$code, msg=$msg, data=$data)"
    }
}
